### PR TITLE
Extended observables

### PR DESCRIPTION
**Pull request by fdmalone**
Refactor of AFQMC estimators to allow for simpler additions of new observables which are mostly based on the (back propagated) density matrices.

### CODE REVIEW

File: src/AFQMC/Estimators/EstimatorHandler.h [hunk 1]
```diff
@@ -10,7 +10,8 @@
 #include "AFQMC/Estimators/EnergyEstimator.h"
 #include "AFQMC/Estimators/BasicEstimator.h"
 #include "AFQMC/Estimators/MixedRDMEstimator.h"
-#include "AFQMC/Estimators/BackPropagatedEstimator.h"
+#include "AFQMC/Estimators/BackPropagatedEstimator.hpp"
+#include "AFQMC/Estimators/newBackPropagatedEstimator.hpp"
```
Comment (fdmalone) [inline src/AFQMC/Estimators/EstimatorHandler.h]: @mmorale3 What is the newBackPropagatedEstimator?

File: src/AFQMC/Estimators/newBackPropagatedEstimator.hpp [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+  using stdCVector_ref = boost::multi::array_ref<ComplexType,1>;
+  using stdCVector = boost::multi::array<ComplexType,1>;
+  using stdCMatrix = boost::multi::array<ComplexType,2>;
+  using stdCTensor = boost::multi::array<ComplexType,3>;
+  using mpi3CVector = boost::multi::array<ComplexType,1,shared_allocator<ComplexType>>;
+  using mpi3CMatrix = boost::multi::array<ComplexType,2,shared_allocator<ComplexType>>;
+  using mpi3CTensor = boost::multi::array<ComplexType,3,shared_allocator<ComplexType>>;
+
+  public:
+
```
[the remark this answers is not in the document]
Comment (fdmalone) [reply src/AFQMC/Estimators/newBackPropagatedEstimator.hpp]: I see it's in notes. I think the source should have more explanation at a top level.

File: src/AFQMC/Estimators/Observables/full1rdm.hpp [hunk 3]
```diff
@@ -0,0 +1,285 @@
+//////////////////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (fdmalone) [reply src/AFQMC/Estimators/Observables/full1rdm.hpp]: This estimator computes the walker averaged one rdm for a given set of pairs of left and right wavefunctions.

File: src/AFQMC/Estimators/EstimatorHandler.h [hunk 4]
```diff
@@ -136,6 +137,9 @@ class EstimatorHandler: public AFQMCInfo
           } else if (name == "back_propagation") {
             estimators.emplace_back(static_cast<EstimPtr>(std::make_shared<BackPropagatedEstimator>(TGgen.getTG(1),info,title,cur,walker_type,wset,*wfn,prop0,impsamp)));
             hdf_output = true;
+          } else if (name == "new_back_propagation") {
+            estimators.emplace_back(static_cast<EstimPtr>(std::make_shared<BackPropagatedEstimator_>(TGgen.getTG(1),info,title,cur,walker_type,wset,*wfn,prop0,impsamp)));
```
Comment (fdmalone) [inline src/AFQMC/Estimators/EstimatorHandler.h]: Is this a temporary thing? Will all estimators go through the new handler eventually?

File: src/AFQMC/Estimators/FullObsHandler.hpp [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+          Xw[iw] = CIcoeff * Ov[iw] * detR[iw][iref] * detR[iw][iref]; 
+      } else if(walker_type == COLLINEAR) {
+        for(int iw=0; iw<nw; iw++) 
+          Xw[iw] = CIcoeff * Ov[iw] * detR[iw][2*iref] * detR[iw][2*iref+1]; 
+      } else if(walker_type == NONCOLLINEAR) {
+        for(int iw=0; iw<nw; iw++) 
+          Xw[iw] = CIcoeff * Ov[iw] * detR[iw][iref]; 
+      } 
+
+      //3. accumulate references
```
Comment (fdmalone) [inline src/AFQMC/Estimators/FullObsHandler.hpp]: What is a reference in this context?

File: src/AFQMC/Estimators/WalkerAveragedObservables.hpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
+//////////////////////////////////////////////////////////////////////
```
Comment (fdmalone) [inline src/AFQMC/Estimators/WalkerAveragedObservables.hpp]: Is this class used anywhere?

File: src/AFQMC/Estimators/newBackPropagatedEstimator.hpp [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+  int max_nback_prop;
+  std::vector<int> nback_prop_steps;  
+
+  FullObsHandler observ0; 
+
+  Wavefunction& wfn0;
+
+  Propagator& prop0;
+
+  // The first element of data stores the denominator of the estimator (i.e., the total
```
Comment (fdmalone) [inline src/AFQMC/Estimators/newBackPropagatedEstimator.hpp]: This comment needs to be deleted at some point.

File: src/AFQMC/Estimators/EstimatorHandler.h (hunk 4 above)
Comment (jtkrogel) [reply src/AFQMC/Estimators/EstimatorHandler.h]: Just a comment: I've observed that use of the word "new" when introducing an updated version of a feature becomes outdated very quickly in retrospect (i.e. usually the "new" thing becomes standard and the original is eventually discarded).

**Comment by mmorale3** [comment]
Hmmm, its the new class I wrote. I forgot to change the name. I left the old class there for comparison and in case any problems appeared. Forgot to change the names before making the commit. I’ll fix it today.

Miguel

File: src/AFQMC/Estimators/EstimatorHandler.h (hunk 4 above)
Comment (prckent) [reply src/AFQMC/Estimators/EstimatorHandler.h]: Agree. These litter the source. Call the new thing by the new algorithm; if it is a replacement for the old algo delete the old implementation.

**Comment by mmorale3** [comment]
All the comments are good. I didn’t clean up the source and also forgot to change class names. I pushed prematurely in order to avoid using code from my private repo for Cornell’s project. I’ll clean it up today before merging.

File: src/AFQMC/Estimators/FullObsHandler.hpp [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+          RefsA.emplace_back(devCMatrix_ref(make_device_ptr(Refs[iw][iref].origin()), 
+                                            {2*NMO,NAEA+NAEB}));
+      }
+      wfn0.DensityMatrix(wset, RefsA, RefsB, G2D, DevOv, false, false, true);
+
+      //2. calculate and accumulate appropriate weights 
+      copy_n( make_device_ptr(DevOv.origin()), nw, Ov.origin());
+      if(walker_type == CLOSED) { 
+        for(int iw=0; iw<nw; iw++) 
+          Xw[iw] = CIcoeff * Ov[iw] * detR[iw][iref] * detR[iw][iref];
```
[the remark this answers is not in the document]
Comment (mmorale3) [reply src/AFQMC/Estimators/FullObsHandler.hpp]: Fixed

File: utils/afqmctools/afqmctools/wavefunction/mol.py [hunk 9]
```diff
@@ -65,7 +65,8 @@ def write_wfn_mol(scf_data, ortho_ao, filename, wfn=None,
                       nelec, norb, verbose=verbose)
     return nelec
 
-def write_qmcpack_wfn(filename, wfn, walker_type, nelec, norb, init=None,
+#def write_qmcpack_wfn(filename, wfn, walker_type, nelec, norb, init=None,
```
[the remark this answers is not in the document]
Comment (mmorale3) [reply utils/afqmctools/afqmctools/wavefunction/mol.py]: mistake, fixed it

[the remark this answers is not in the document]
Comment (mmorale3) [reply utils/afqmctools/afqmctools/wavefunction/mol.py]: well, not really
did you fix the issue with uhf vs walker_type?

File: src/AFQMC/Estimators/EstimatorHandler.h (hunk 4 above)
Comment (mmorale3) [reply src/AFQMC/Estimators/EstimatorHandler.h]: I left the old one for comparison and forgot to remove it/rename the new one.
Fixed it.

File: src/AFQMC/Estimators/newBackPropagatedEstimator.hpp (hunk 7 above)
Comment (mmorale3) [reply src/AFQMC/Estimators/newBackPropagatedEstimator.hpp]: Done

File: src/AFQMC/Estimators/FullObsHandler.hpp (hunk 5 above)
Comment (mmorale3) [reply src/AFQMC/Estimators/FullObsHandler.hpp]: A reference is a term in the (possibly) MSD expansion being back propagated.
Each determinant needs to be back propagated independently and the average
done at the end. References stands for "Reference Determinants" or 
"Determinants in the Trial Wavefunction"

File: src/AFQMC/Estimators/WalkerAveragedObservables.hpp (hunk 6 above)
Comment (mmorale3) [reply src/AFQMC/Estimators/WalkerAveragedObservables.hpp]: nope, just removed it

File: src/AFQMC/Estimators/newBackPropagatedEstimator.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmorale3) [reply src/AFQMC/Estimators/newBackPropagatedEstimator.hpp]: I'll write some temporary notes. Overall the code is still not properly commented

File: src/AFQMC/Estimators/Observables/full1rdm.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmorale3) [reply src/AFQMC/Estimators/Observables/full1rdm.hpp]: For naming consistency it should be full1RDM no? 
where?

[the remark this answers is not in the document]
Comment (fdmalone) [reply src/AFQMC/Estimators/Observables/full1rdm.hpp]: Not very important, camelCase for classes?

File: utils/afqmctools/afqmctools/wavefunction/mol.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fdmalone) [reply utils/afqmctools/afqmctools/wavefunction/mol.py]: Eh clearly not. We can come back to this.

**Comment by ye-luo** [comment]
Okay to test

**Comment by ye-luo** [comment]
Test this please.

**Comment by ye-luo** [comment]
Test this please.

**Pull request merged**